File: gpt-4-turbo/stage2/kernels_construct_default_clause.f90
!T1:kernels construct default clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        INTEGER, DIMENSION(100) :: test_array
        INTEGER :: i
        
        ! Initialize the array
        DO i = 1, SIZE(test_array)
          test_array(i) = i
        END DO

        ! Apply the kernels construct
        !$acc kernels default(present)
        DO i = 1, SIZE(test_array)
          test_array(i) = test_array(i) * 2
        END DO
        !$acc end kernels
        
        ! Check for correctness
        DO i = 1, SIZE(test_array)
          IF (test_array(i) /= i * 2) THEN
            errors = errors + 1
          END IF
        END DO

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        EXTERNAL test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM

File: gpt-4-turbo/stage2/kernels_construct_default_clause.cpp
#include <cstdlib>
#include <iostream>

// Assuming acc_testsuite.h provides necessary OpenACC testing support
// In this case, it's just conceptual.
#include "acc_testsuite.h"

#ifndef T1
//T1:kernels construct default clause,V:2.7-3.3
int test1(){
    int err = 0;
    srand(SEED);

    const int size = 100;
    int a[size];
    for (int i = 0; i < size; ++i) {
        a[i] = rand() % 100; // Assign random numbers
    }

    int scalar = 5; // A scalar variable to test capture by value

    #pragma acc kernels default(present) copy(scalar)
    for (int i = 0; i < size; ++i) {
        a[i] += scalar; // Should add 'scalar' value to each element of 'a'
        scalar = 999; // Try to change scalar inside the kernels region
    }

    // Since 'scalar' should be captured by value, changes inside the
    // kernel region should not affect the original 'scalar' variable.
    if (scalar != 5) {
        err++;
    }

    // Verifying if 'a' array has been correctly modified
    for (int i = 0; i < size; ++i) {
        if (a[i] < 5 || a[i] > 104) { // Since random was from 0 to 99
            err++;
            break;
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}